pkg/volume/volume.go

```diff
@@ -132,6 +132,48 @@ type Unmounter interface {
 	TearDownAt(dir string) error
 }
 
+//
+type BlockVolumeMapper interface {
+	Volume
+	//
+	CanBlockMap() error
+	// SetUpDevice creates a symbolic link to the volume on a
+	// self-determined directory path. This may be called more than
+	// once, so implementations must be idempotent.
+	SetUpDevice(podUID types.UID) error
+	// SetUpDeviceAt a symbolic link to the volume on a specified
+	// directory path, which may or may not exist yet. This may be
+	// called more than once, so implementations must be idempotent.
+	SetUpDeviceAt(dir string, podUID types.UID) error
+	// GetDevicePath obtains a block device path on the node.
+	GetDevicePath() (string, error)
+	// GetGlobalMapPath generates a global map path to create
+	// a symbolic link to a block device.
+	GetGlobalMapPath(spec *Spec) (string, error)
```

Comment:
Why does the plugin need to implement these two methods? Shouldn't volume manager be able to generate the paths without plugin help?

Reply:
Global map path contains plugin own unique directory name like below.
`28000001ff0414e2-lun-0 => WWN + lun number` 
This is similar to current filesystem case.
Therefore, volume manager can't generate this path by itself. If we use another plugin independent unique volume information for this path(ex volume name?), plugin doesn't need to implement this method.

```
# find plugins | ag devices 
plugins/kubernetes.io/fc/volumeDevices
plugins/kubernetes.io/fc/volumeDevices/28000001ff0414e2-lun-0
plugins/kubernetes.io/fc/volumeDevices/28000001ff0414e2-lun-0/4c65e761-a264-11e7-bf52-525400f59db4
```

GetPodDeviceMapPath() returns similar value to GetPath() in each plugin.
Volume manager might generate this value using volume spec. 

```
func (fc *fcDisk) GetPath() string {
        name := fcPluginName
        // safe to use PodVolumeDir now: volume teardown occurs before pod is cleaned up
        return fc.plugin.host.GetPodVolumeDir(fc.podUID, utilstrings.EscapeQualifiedNameForDisk(name), fc.volName)
}
```

Pod device map path
```
# find pods | ag devices
pods/4c65e761-a264-11e7-bf52-525400f59db4/volumeDevices
pods/4c65e761-a264-11e7-bf52-525400f59db4/volumeDevices/kubernetes.io~fc
pods/4c65e761-a264-11e7-bf52-525400f59db4/volumeDevices/kubernetes.io~fc/pv0001
pods/4c65e761-a264-11e7-bf52-525400f59db4/volumeDevices/kubernetes.io~fc/pv0001/symlink
```

Reply:
Ah thanks for the examples, it helps a lot.  Can you add comments with the format of the various paths?